crypto/rsa/rsa_sign.c

```diff
@@ -16,60 +16,189 @@
 #include "internal/x509_int.h"
 #include "rsa_locl.h"
 
+/* SHA OIDs: (2 16 840 1 101 3 4 2 n), with and without NULL parameters */
+#define SHA_DATA(name, n, sz) \
+static const unsigned char id_##name##_oid[] = { \
+    0x30, 0x11 + sz, \
+      0x30, 0x0d, \
+        0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, n, \
+        0x05, 0x00, \
+      0x04, sz \
+}; \
+static const unsigned char id_##name##_oid_nn[] = { \
+    0x30, 0x0f + sz, \
+      0x30, 0x0b, \
+        0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, n, \
+      0x04, sz \
+}
+
+/* MD2 and MD5: (1 2 840 113549 2 n), parameters always present (NULL) */
+#define MD_DATA(name, n, sz) \
+static const unsigned char id_##name##_oid[] = { \
+    0x30, 0x10 + sz, \
+      0x30, 0x0c, \
+        0x06, 0x08, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x02, n, \
+        0x05, 0x00, \
+      0x04, sz \
+}
+
+#define MD_CASE(name) \
+    case NID_##name: \
+        *len = sizeof(id_##name##_oid); \
+        return id_##name##_oid
+#define MD_CASE_NN(name) \
+    case NID_##name: \
+        *len = sizeof(id_##name##_oid_nn); \
+        return id_##name##_oid_nn
+
+#ifndef FIPS_MODE
+MD_DATA(md2, 0x02, 0x10);
+MD_DATA(md5, 0x05, 0x10);
+
+/* SHA-1 (1 3 14 3 2 26) */
+static const unsigned char id_sha1_oid[] = {
+    0x30, 0x21, 0x30, 0x09, 0x06, 0x05, 0x2b, 0x0e, 0x03, 0x02, 0x1a, 0x05,
+    0x00, 0x04, 0x14
+};
+
+/* SHA-1 (1 3 14 3 2 26), no NULL parameters */
+static const unsigned char id_sha1_oid_nn[] = {
+    0x30, 0x1f, 0x30, 0x07, 0x06, 0x05, 0x2b, 0x0e, 0x03, 0x02, 0x1a, 0x04,
+    0x14
+};
+
+/* MDC-2 (2 5 8 3 101) */
+static const unsigned char id_mdc2_oid[] = {
+    0x30, 0x1c, 0x30, 0x08, 0x06, 0x04, 0x55, 0x08, 0x03, 0x65, 0x05, 0x00,
+    0x04, 0x10
+};
+#endif
+
+SHA_DATA(sha256, 0x01, 0x20);
```

Comment:
Perhaps replace the last number with `SHA256_DIGEST_LENGTH`? Similarly with the others below?

Reply:
Done.
There are no specific constants for SHA3 or the truncated SHA2 digests so I just reused the SHA2 ones.